music21/stream/base.py

```diff
@@ -8988,46 +9013,58 @@ def bestMatch(target, divisors):
         # quarterLengthMin = quarterLengthDivisors[0]
 
-        # TODO: Use new filters...
         if inPlace is False:
             returnStream = self.coreCopyAsDerivation('quantize')
-
         else:
             returnStream = self
 
         useStreams = [returnStream]
         if recurse is True:
-            for obj in returnStream.recurse():
-                if 'Stream' in obj.classes:
-                    useStreams.append(obj)
+            useStreams = returnStream.recurse(streamsOnly=True, includeSelf=True)
 
         for useStream in useStreams:
-            for e in useStream._elements:
+            for i, e in enumerate(useStream._elements):
                 if processOffsets:
                     o = useStream.elementOffset(e)
                     sign = 1
                     if o < 0:
                         sign = -1
                         o = -1 * o
-                    unused_error, oNew, signedError = bestMatch(float(o), quarterLengthDivisors)
+                    unused_error, oNew, signedError, unused_inverse_divisor = bestMatch(
+                        float(o), quarterLengthDivisors)
                     useStream.coreSetElementOffset(e, oNew * sign)
                     if (hasattr(e, 'editorial')
                             and signedError != 0):
                         e.editorial.offsetQuantizationError = signedError * sign
                 if processDurations:
                     ql = e.duration.quarterLength
                     ql = max(ql, 0)  # negative ql possible in buggy MIDI files?
-                    unused_error, qlNew, signedError = bestMatch(
+                    unused_error, qlNew, signedError, unused_inverse_divisor = bestMatch(
                         float(ql), quarterLengthDivisors)
+                    # Check that any gaps from this quantized duration to the next onset
+                    # are at least as large as the smallest quantization unit (largest divisor)
+                    # If not, then re-quantize this duration with the divisor
+                    # that will be used to quantize the next element's offset
+                    if processOffsets and i + 1 < len(useStream._elements):
+                        next_element = useStream._elements[i + 1]
+                        next_offset = useStream.elementOffset(next_element)
+                        look_ahead_result = bestMatch(float(next_offset), quarterLengthDivisors)
+                        next_offset = look_ahead_result[1]
+                        next_divisor = 1 / look_ahead_result[3]
+                        if 0 < next_offset - (e.offset + qlNew) < 1 / max(quarterLengthDivisors):
+                            unused_error, qlNew, signedError, unused_inverse_divisor = bestMatch(
+                                float(ql), (next_divisor,))
                     # Enforce nonzero duration for non-grace notes
+                    # TODO: what about rests? remove rests with ql=0?
```

Comment:
PLEASE!

Reply:
(I was going to say, remove all zero-length notes, but I suppose all zero-length notes should become part of a Chord with the next note, etc.  but that's too much to ask here).

Reply:
in the next block, all `GeneralNotes` are given a minimum nonzero duration (usually 0.25)-- believe that was #641. 

But my idea here was that's too aggressive when it comes to rests: we should let 0.0 rests that come out of quantization just get deleted instead of beefing them up. Since this is mostly used with MIDI and MIDI doesn't have a notated absence that should be preserved, it just has ... absences. Shouldn't be an effortful addition.

Reply:
Ah, but this would probably only happen when either quantizing a stream with notated rests already in it, since in MIDI parsing quantizing comes before rest creation.